Could not pass event PlayerInteractEvent to XClaim v1.10.5

Im using stock SPIGOT. Tried 1.20.4 and 1.20.2 versions, both have this issue while someone has opened his claiming interface console gets spammed with these messages constantly.


[23:06:57] [Server thread/ERROR]: Could not pass event PlayerInteractEvent to XClaim v1.10.5
org.bukkit.event.EventException: null
        at org.bukkit.plugin.java.JavaPluginLoader$1.execute(JavaPluginLoader.java:310) ~[spigot-api-1.20.2-R0.1-SNAPSHOT.jar:?]
        at org.bukkit.plugin.RegisteredListener.callEvent(RegisteredListener.java:70) ~[spigot-api-1.20.2-R0.1-SNAPSHOT.jar:?]
        at org.bukkit.plugin.SimplePluginManager.fireEvent(SimplePluginManager.java:589) ~[spigot-api-1.20.2-R0.1-SNAPSHOT.jar:?]
        at org.bukkit.plugin.SimplePluginManager.callEvent(SimplePluginManager.java:576) ~[spigot-api-1.20.2-R0.1-SNAPSHOT.jar:?]
        at org.bukkit.craftbukkit.v1_20_R2.event.CraftEventFactory.callPlayerInteractEvent(CraftEventFactory.java:570) ~[spigot-1.20.2-R0.1-SNAPSHOT.jar:3950-Spigot-2888990-d485cde]
        at org.bukkit.craftbukkit.v1_20_R2.event.CraftEventFactory.callPlayerInteractEvent(CraftEventFactory.java:532) ~[spigot-1.20.2-R0.1-SNAPSHOT.jar:3950-Spigot-2888990-d485cde]
        at net.minecraft.server.level.PlayerInteractManager.a(PlayerInteractManager.java:187) ~[spigot-1.20.2-R0.1-SNAPSHOT.jar:3950-Spigot-2888990-d485cde]
        at net.minecraft.server.network.PlayerConnection.a(PlayerConnection.java:1484) ~[spigot-1.20.2-R0.1-SNAPSHOT.jar:3950-Spigot-2888990-d485cde]
        at net.minecraft.network.protocol.game.PacketPlayInBlockDig.a(SourceFile:42) ~[spigot-1.20.2-R0.1-SNAPSHOT.jar:3950-Spigot-2888990-d485cde]
        at net.minecraft.network.protocol.game.PacketPlayInBlockDig.a(SourceFile:8) ~[spigot-1.20.2-R0.1-SNAPSHOT.jar:3950-Spigot-2888990-d485cde]
        at net.minecraft.network.protocol.PlayerConnectionUtils.lambda$0(PlayerConnectionUtils.java:32) ~[spigot-1.20.2-R0.1-SNAPSHOT.jar:3950-Spigot-2888990-d485cde]
        at net.minecraft.server.TickTask.run(SourceFile:18) ~[spigot-1.20.2-R0.1-SNAPSHOT.jar:3950-Spigot-2888990-d485cde]
        at net.minecraft.util.thread.IAsyncTaskHandler.d(SourceFile:156) ~[spigot-1.20.2-R0.1-SNAPSHOT.jar:3950-Spigot-2888990-d485cde]
        at net.minecraft.util.thread.IAsyncTaskHandlerReentrant.d(SourceFile:23) ~[spigot-1.20.2-R0.1-SNAPSHOT.jar:3950-Spigot-2888990-d485cde]
        at net.minecraft.server.MinecraftServer.b(MinecraftServer.java:1133) ~[spigot-1.20.2-R0.1-SNAPSHOT.jar:3950-Spigot-2888990-d485cde]
        at net.minecraft.server.MinecraftServer.d(MinecraftServer.java:1) ~[spigot-1.20.2-R0.1-SNAPSHOT.jar:3950-Spigot-2888990-d485cde]
        at net.minecraft.util.thread.IAsyncTaskHandler.x(SourceFile:130) ~[spigot-1.20.2-R0.1-SNAPSHOT.jar:3950-Spigot-2888990-d485cde]
        at net.minecraft.server.MinecraftServer.bg(MinecraftServer.java:1112) ~[spigot-1.20.2-R0.1-SNAPSHOT.jar:3950-Spigot-2888990-d485cde]
        at net.minecraft.server.MinecraftServer.x(MinecraftServer.java:1105) ~[spigot-1.20.2-R0.1-SNAPSHOT.jar:3950-Spigot-2888990-d485cde]
        at net.minecraft.util.thread.IAsyncTaskHandler.bp(SourceFile:115) ~[spigot-1.20.2-R0.1-SNAPSHOT.jar:3950-Spigot-2888990-d485cde]
        at net.minecraft.server.MinecraftServer.u_(MinecraftServer.java:1088) ~[spigot-1.20.2-R0.1-SNAPSHOT.jar:3950-Spigot-2888990-d485cde]
        at net.minecraft.server.MinecraftServer.w(MinecraftServer.java:1000) ~[spigot-1.20.2-R0.1-SNAPSHOT.jar:3950-Spigot-2888990-d485cde]
        at net.minecraft.server.MinecraftServer.lambda$0(MinecraftServer.java:298) ~[spigot-1.20.2-R0.1-SNAPSHOT.jar:3950-Spigot-2888990-d485cde]
        at java.lang.Thread.run(Thread.java:833) ~[?:?]
Caused by: java.util.ConcurrentModificationException
        at java.util.HashMap$HashIterator.nextNode(HashMap.java:1597) ~[?:?]
        at java.util.HashMap$KeyIterator.next(HashMap.java:1620) ~[?:?]
        at codes.wasabi.xclaim.api.Claim.addChunk(Claim.java:352) ~[?:?]
        at codes.wasabi.xclaim.gui.ChunkEditor$Events.onInteract(ChunkEditor.java:227) ~[?:?]
        at jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104) ~[?:?]
        at java.lang.reflect.Method.invoke(Method.java:577) ~[?:?]
        at org.bukkit.plugin.java.JavaPluginLoader$1.execute(JavaPluginLoader.java:306) ~[spigot-api-1.20.2-R0.1-SNAPSHOT.jar:?]
        ... 23 more

--- a/src/main/java/codes/wasabi/xclaim/api/Claim.java
+++ b/src/main/java/codes/wasabi/xclaim/api/Claim.java
@@ -31,7 +31,7 @@ import java.util.stream.Stream;
 
 public class Claim {
 
-    private static final Set<Claim> registry = new HashSet<>();
+    private static final Set<Claim> registry = Collections.synchronizedSet(new HashSet<>());
     private static final ReentrantReadWriteLock registryLock = new ReentrantReadWriteLock();
 
     public static @NotNull @UnmodifiableView Set<Claim> getAll() {
@@ -272,19 +272,24 @@ public class Claim {
         String targetName = this.name;
         int i = 0;
         UUID ownerId = this.owner.getUniqueId();
-        while (true) {
-            boolean ok = true;
-            for (Claim c : registry) {
-                if (c == this) continue;
-                if (c.owner == null || c.owner.getUniqueId() != ownerId) continue;
-                if (c.name.equalsIgnoreCase(targetName)) {
-                    i++;
-                    targetName = desiredName + " (" + i + ")";
-                    ok = false;
-                    break;
+        registryLock.readLock().lock();
+        try {
+            while (true) {
+                boolean ok = true;
+                for (Claim c : registry) {
+                    if (c == this) continue;
+                    if (c.owner == null || c.owner.getUniqueId() != ownerId) continue;
+                    if (c.name.equalsIgnoreCase(targetName)) {
+                        i++;
+                        targetName = desiredName + " (" + i + ")";
+                        ok = false;
+                        break;
+                    }
                 }
+                if (ok) break;
             }
-            if (ok) break;
+        } finally {
+            registryLock.readLock().unlock();
         }
         this.name = targetName;
     }
@@ -349,9 +354,14 @@ public class Claim {
         if (ret) {
             generateBounds();
             if (manageHandlers) {
-                for (Claim c : registry) {
-                    if (c == this) continue;
-                    if (c.chunks.contains(ref)) c.removeChunk(chunk);
+                registryLock.readLock().lock();
+                try {
+                    for (Claim c : registry) {
+                        if (c == this) continue;
+                        if (c.chunks.contains(ref)) c.removeChunk(chunk);
+                    }
+                } finally {
+                    registryLock.readLock().unlock();
                 }
             } else if (claim) {
                 claim();

--- a/src/main/java/codes/wasabi/xclaim/api/Claim.java
+++ b/src/main/java/codes/wasabi/xclaim/api/Claim.java
@@ -31,7 +31,7 @@ import java.util.stream.Stream;
 
 public class Claim {
 
-    private static final Set<Claim> registry = Collections.synchronizedSet(new HashSet<>());
+    private static final Set<Claim> registry = new HashSet<>();
     private static final ReentrantReadWriteLock registryLock = new ReentrantReadWriteLock();
 
     public static @NotNull @UnmodifiableView Set<Claim> getAll() {

